functions/wagers.js

```diff
@@ -0,0 +1,321 @@
+const {v4: uuidv4} = require('uuid');
+const {notifyGroupOfWager} = require('./groups');
+
+const db = admin.firestore();
+
+const ActionTypes = {
+    WIN   : 'WIN',
+    LOSS  : 'LOSS',
+    PUSH  : 'PUSH',
+    PAID  : 'PAID',
+    CANCEL: 'CANCEL',
+    CONFIRM_CANCEL : 'CONFIRM_CANCEL',
+    CONFIRM_WINNER : 'CONFIRM_WINNER',
+};
+
+const Statuses = {
+    PAID    : 'paid',
+    OPEN    : 'open',
+    BOOKED  : 'booked',
+    PENDING : 'pending',
+    PROPOSED: 'resolutionProposed',
+    REJECTED: 'rejected',
+    CANCEL_PROPOSED : 'cancellationProposed',
+}
+
+const fail = (message) => {
+    throw new functions.https.HttpsError('failed-precondition', message)
+}
+
+const onWagerWrite = async (change) => {
+    const wager = change.after.data();
+    if (wager && wager.proposedTo && wager.proposedTo.uid) {
+
+        if (wager.status === Statuses.PENDING) {
+            const snapshot = await db.collection('users').doc(wager.proposedTo.uid).get()
+            const proposedToUser = snapshot.data();
+
+            const mail = {
+                to: proposedToUser.email,
+                template: {
+                    name: 'new-wager',
+                    data: {
+                        proposedBy: wager.proposedBy.displayName
+                    }
+                }
+            }
+
+            await db.collection('mail').add(mail);
+        }
+    }
+}
+
+async function lookupWager(groupId, wagerId) {
+    const doc = await db.collection('groups').doc(groupId)
+        .collection('wagers').doc(wagerId).get();
+
+    if (!doc.exists) {
+        throw new functions.https.HttpsError('failed-precondition', 'This wager doesn\'t exist');
+    }
+    return doc.data();
+}
+
+function wagerIncludesUser(wager, uid) {
+    return wager.proposedTo.uid === uid 
+        || wager.proposedBy.uid === context.auth.uid;
+}
+
+function handlePaidAction(wager, user) {
+    if(!wager.winner){
+        fail('A wager must have a winner before you pay it out');
+    }
+
+    if(wager.resolutionProposedBy.uid === user.uid){
+        fail('You were the one to propose a resolution');
+    }
+
+    if(wager.status !== Statuses.PROPOSED){
+        fail('Wager must be resolved before paying out');
+    }
+
+    return {
+        ...wager,
+        status: Statuses.PAID,
+    }
+}
+
+function handleProposeResolution(wager, actionType, user, opponent) {
+    if(wager.status !== Statuses.BOOKED){
```

Comment:
Adding PUSH to the same resolution as win/loss means this check is made for push as well.